README.md
# Happy Feet

Application state management with triggers that span both urgent and non-urgent states to account for more graceful and staggered recoveries.


## State Management

Sometimes healthy isn't so straight forward. Intermediate states can help fill the `void`.

```js
const happy = require('happy-feet')();

// by default, happy.state === happy.STATE.HAPPY

// sometimes apps want to differentiate between unhealthy, and not-yet-healthy...
happy.state = happy.STATE.STARTING;

// app is now running
happy.state = happy.STATE.HAPPY;

// something went wrong, but we're still operational.
// eventual escalation to UNHAPPY
happy.state = happy.STATE.WARN;

// other times, maybe you just want your own custom state -- who are we to judge?
happy.state = 'BLACK_HOLE';

// DANGER WILL ROBINSON!
happy.state = happy.STATE.UNHAPPY;

// once state set to UNHAPPY it is irreversible, future states will be ignored
```


## Usage

The core functionality of `happy-feet` lies in setting the `state` property of
your `happy-feet` instance. This state may be set automatically by `happy-feet`
as it monitors your process or set by your application. The value of this state
can be automatically reflected at a web service URL via one of the provided 
middlewares. This can be used to signal to supervising processes that are 
monitoring this health indicator when they need to take corrective action.

You may also decide to manually use the `happy-feet` API to roll your own state
management logic:

```js
const happy = require('happy-feet')({ /* optional options */ });

// happy.state === happy.STATE.HAPPY by default

if (happy.state !== happy.STATE.HAPPY) {
  // do something about it
}
```

There are three built-in values that have a special meaning to `happy-feet`:

| State | Meaning |
|-------|---------|
| `happy.STATE.HAPPY`   | This indicates that your service is in a healthy state |
| `happy.STATE.WARN`    | This means your service is having problems. The health check page will reflect an OK status, but eventually the state will automatically transition to `UNHAPPY`. |
| `happy.STATE.UNHAPPY` | This signals that your service instance should be considered ready to be terminated/restarted due to being in a bad state. |

You can also use custom states if you're rolling your own health status
implementation.

Automatic state change triggers include soft and hard limits. Soft limits trigger
a warning state (a temporary successful state) indicating the eventual need to be restarted,
while hard limits trigger a state of immediate urgency. 

The automatic transition from the `WARN` state to `UNHAPPY` is done after a
random amount of time, which is configurable with the `escalationSoftLimitMin`
and `escalationSoftLimitMax` settings. The purpose of this randomness is to
avoid each member of a service cluster being taken down at the same time in case
of full cluster impacting faults.

Here's a full list of configuration options for `happy-feet`:

| Option | Type | Default | Info |
| --- | --- | --- | --- |
| escalationSoftLimitMin | `number` | `60` | Minimum time (in seconds) before a `WARN` state may be escalated to an `UNHAPPY` state. |
| escalationSoftLimitMax | `number` | `600` | Maximum time (in seconds) before a `WARN` state may be escalated to an `UNHAPPY` state. |
| uncaughtExceptionSoftLimit | `number` | `1` | Number of uncaught exceptions before `WARN` state. |
| uncaughtExceptionHardLimit | `number` | `undefined` | Number of uncaught exceptions before `UNHAPPY` state. Disabled by default. |
| unhandledRejectionSoftLimit | `number` | `undefined` | Number of unhandled rejections before `WARN` state. Disabled by default. |
| unhandledRejectionHardLimit | `number` | `undefined` | Number of unhandled rejections before `UNHAPPY` state. Disabled by default. |
| rssSoftLimit | `number` | `undefined` | Memory Resident Set Size (in bytes) before `WARN` state. Disabled by default. |
| rssHardLimit | `number` | `undefined` | Memory Resident Set Size (in bytes) before `UNHAPPY` state. Disabled by default. |
| heapSoftLimit | `number` | `undefined` | Total heap size (in bytes) before `WARN` state. Disabled by default. |
| heapHardLimit | `number` | `undefined` | Total heap size (in bytes) before `UNHAPPY` state. Disabled by default. |
| eventLoopSoftLimit | `number` | `undefined` | Event Loop delay (in ms) before `WARN` state. Disabled by default. Recommended value of `150` or higher. |
| eventLoopHardLimit | `number` | `undefined` | Event Loop delay (in ms) before `UNHAPPY` state. Disabled by default. Recommended value of `500` or higher. |
| timeLimitMin | `number` | `undefined` | Minimum time (in seconds) before `UNHAPPY` state. Disabled by default. Useful for periodic application restarts. Both `timeLimitMin` and `timeLimitMax` must be set to use this feature. |
| timeLimitMax | `number` | `undefined` | Maximum time (in seconds) before `UNHAPPY` state. Disabled by default. Useful for periodic application restarts. Both `timeLimitMin` and `timeLimitMax` must be set to use this feature. |
| logger | `{ warn,error }` | `console` | Logging interface to use when state changes occur. Defaults to use `console`. |
| gracePeriod | `number` | `300` (5 mins) | The time (in seconds) before any thresholds can trigger an unhealthy state change to alleviate startup pains |
| logOnUnhappy | `boolean` | `true` | If enabled, all *checks* for state will be logged if NOT `HAPPY` to help troubleshoot state changes |

A `happy-feet` instance has this interface:

| Property | Type | Info |
| --- | --- | --- |
| state | `string` | Get/set the current state of your process. |
| updateState(state, reason, code) | `function` | Calling this does the same thing as setting the `state` property, except it allows you to log a `reason` for the change. The code argument is there to enable identifying the reason for transitions programmatically (see events below). |
| destroy() | `function` | Use this to free the instance. |

Additionally, each `happy-feet` instance is an `EventEmitter`. It emits a `change` event with the following parameters whenever its state changes:

| Parameter | Type | Info |
|-----------|------|------|
| `state`   | `string` | The new state |
| `reason`  | `string` | The textual description of the change |
| `code`    | `string` | Can be any value for custom implementations or `manual` if the state was changed through assignment. For automatic transitions can be one of the following: `uncaughtExceptions`, `unhandledRejections`, `memory`, `eventLoop`, or `escalation` |

## Connect Usage

If you've already got an (Connect) API, attach a handler like so:
```
const happyConnect = require('happy-feet/connect');
const handler = happyConnect({ /* options */ }, { /* optional happy options */ });

app.use(handler);

// Manually change the state
handler.happy.state = handler.happy.STATE.WARN; 
```

| Option | Type | Default | Info |
| --- | --- | --- | --- |
| url | `string` | `"/_health"` | Route to return health status. Defaults to common Kubernetes healthcheck route. |
| method | `string` | `"GET"` | Method required to return health status. |
| errorStatus | `number` | `500` | Status code returned if in `UNHAPPY` state. |
| status | `object` | `{}` | A collection of custom overrides for status responses based on individual states. |
| status[STATE].statusCode | `number` | `200|500` | Status code returned for the given state. By default `${errorStatus}` is returned for `UNHAPPY` state, otherwise `200`.  |
| status[STATE].body | `string` | `${STATE}` | Body message returned for the given state. By default `${STATE}` will be returned verbatim. |
| status[STATE].contentType | `string` | `text/plain` | Content type to respond with. |

| Return Property | Type | Info |
| --- | --- | --- |
| happy | `Happy` | Instance of [Happy](#usage). |


## Express Usage

If you've already got an (Express) API, attach a handler like so:
```
const happyExpress = require('happy-feet/express');

const handler = happyExpress({ /* options */ }, { /* optional happy options */ });
app.get('/_health', handler);

// Manually change the state
handler.happy.state = handler.happy.STATE.WARN; 
```

| Option | Type | Default | Info |
| --- | --- | --- | --- |
| errorStatus | `number` | `500` | Status code returned if in `UNHAPPY` state. |
| status | `object` | `{}` | A collection of custom overrides for status responses based on individual states. |
| status[STATE].statusCode | `number` | `200|500` | Status code returned for the given state. By default `${errorStatus}` is returned for `UNHAPPY` state, otherwise `200`.  |
| status[STATE].body | `string` | `${STATE}` | Body message returned for the given state. By default `${STATE}` will be returned verbatim. |
| status[STATE].contentType | `string` | `text/plain` | Content type to respond with. |

| Return Property | Type | Info |
| --- | --- | --- |
| happy | `Happy` | Instance of [Happy](#usage). |


## API Usage

Or if your service does not expose an API, you can use this helper to expose your healthcheck for you.

```
const happyApi = require('happy-feet/api');

const handler = happyApi({ /* options */ }, { /* optional happy options */ });
```

| Option | Type | Default | Info |
| --- | --- | --- | --- |
| port | 'number' | `80` | HTTP port to bind to. |
| url | `string` | `"/_health"` | Route to return health status. Defaults to common Kubernetes healthcheck route. |
| method | `string` | `"GET"` | Method required to return health status. |
| errorStatus | `number` | `500` | Status code returned if in `UNHAPPY` state. |
| status | `object` | `{}` | A collection of custom overrides for status responses based on individual states. |
| status[STATE].statusCode | `number` | `200|500` | Status code returned for the given state. By default `${errorStatus}` is returned for `UNHAPPY` state, otherwise `200`.  |
| status[STATE].body | `string` | `${STATE}` | Body message returned for the given state. By default `${STATE}` will be returned verbatim. |
| status[STATE].contentType | `string` | `text/plan` | Content type to responde with. |

| Return Property | Type | Info |
| --- | --- | --- |
| happy | `Happy` | Instance of [Happy](#usage). |
| server | `http.Server` | Instance of [http.Server](https://nodejs.org/api/http.html#http_class_http_server). |


## Advanced Healthchecks

The main advantage of *Advanced Healthchecks* over the default escalation
threshold is that a centralized system can coordinate restarts in a more
graceful and controlled manner without impact to availability.

If your system for monitoring healthchecks is capable of discerning between
catastrophic (remove from LB immediately) and unhealthy (some undesirable event
or threshold, operational but eventually should be restarted), `happy.STATE.WARN`
(HTTP `statusMessage` of `WARN`) can be leveraged to gracefully handle rolling
restarts in whatever fashion deemed acceptable. By default, `WARN` states will be
automatically escalated to `UNHAPPY` after a period of time to avoid vulnerabilities
resulting in mass concurrent crashes and ultimately impact to customers and
availability of your services.


CHANGELOG.md
# CHANGELOG


## 2.2.0

* ***ENHANCEMENT*** Add TypeScript declarations


## 2.1.0

* ***ENHANCEMENT*** Support for monitoring total heap usage via `heapSoftLimit` and `heapHardLimit`.


## 2.0.1

* ***FIX*** Do not emit the `change` event more than once for the same event.


## 2.0

* ***ENHANCEMENT*** Addition of `gracePeriod` which will prevent state transitions
  during that initial grace period, which defaults to 5 minutes. This is intended
  help protect newly launched services from causing cascading failures.
* ***ENHANCEMENT*** Default thresholds for transitioning into an `UNHAPPY` state
  have been increased to be less aggressive, playing nice with autoscalers.


lib/core/Happy.js
const { EventEmitter } = require('events');
const states = require('./states');

const codes = {
  uncaughtExceptions: 'uncaughtExceptions',
  unhandledRejections: 'unhandledRejections',
  memory: 'memory',
  eventLoop: 'eventLoop',
  escalation: 'escalation',
  timeLimit: 'timeLimit',
  manual: 'manual'
};

module.exports = class Happy extends EventEmitter {
  constructor(opts) {
    super();
    this.options = Object.assign({
      escalationSoftLimitMin: 60, // 1min
      escalationSoftLimitMax: 600, // 10min
      uncaughtExceptionSoftLimit: 1,
      logger: console,
      gracePeriod: 60 * 5, // 5 min grace period
      logOnUnhappy: true
    }, opts);
    this._state = states.HAPPY;
    this.startTime = Date.now();
    this.uncaughtExceptions = 0;
    this.unhandledRejections = 0;
    this.eventLoop = 0;
    // determine escalation limit up front
    this.escalationSoftLimit = Math.round(Math.random() * (this.options.escalationSoftLimitMax - this.options.escalationSoftLimitMin) + this.options.escalationSoftLimitMin) * 1000;

    if (this.options.uncaughtExceptionSoftLimit || this.options.uncaughtExceptionHardLimit) {
      this.onUncaughtException = ex => {
        this.uncaughtExceptions++;
        this.options.logger && this.options.logger.error('onUncaughtException:', ex.stack);
      };
      process.on('uncaughtException', this.onUncaughtException);
    }

    if (this.options.unhandledRejectionSoftLimit || this.options.unhandledRejectionHardLimit) {
      this.onUnhandledRejection = ex => {
        this.unhandledRejections++;
        this.options.logger && this.options.logger.error('onUnhandledRejection:', ex.stack);
      };
      process.on('unhandledRejection', this.onUnhandledRejection);
    }

    if (this.options.timeLimitMin && this.options.timeLimitMax) {
      const timeLimit = this.options.timeLimitMin + Math.round(Math.random() * (this.options.timeLimitMax - this.options.timeLimitMin));
      this.timeLimitTimer = setTimeout(() => {
        this.updateState(states.UNHAPPY, `timeLimit of ${this.options.timeLimitMin} exceeded`, codes.timeLimit);
      }, timeLimit * 1000 /* ms/s */).unref();
    }

    if (this.options.eventLoopSoftLimit || this.options.eventLoopHardLimit) {
      // based on https://github.com/STRML/node-toobusy/blob/master/toobusy.js
      const interval = 250;
      const smoothingFactor = 1 / 2;
      let lastTime = Date.now();
      this.eventLoopTimer = setInterval(() => {
        const now = Date.now();
        let lag = now - lastTime;
        lag = Math.max(0, lag - interval);
        this.eventLoop = (smoothingFactor * lag) + ((1 - smoothingFactor) * this.eventLoop);
        lastTime = now;
      }, interval).unref();
    }
  }

  destroy() {
    if (this.timeLimitTimer) {
      clearTimeout(this.timeLimitTimer);
      this.timeLimitTimer = undefined;
    }

    if (this.eventLoopTimer) {
      clearInterval(this.eventLoopTimer);
      this.eventLoopTimer = undefined;
    }

    if (this.onUncaughtException) {
      process.removeListener('uncaughtException', this.onUncaughtException);
      this.onUncaughtException = undefined;
    }

    if (this.onUnhandledRejection) {
      process.removeListener('unhandledRejection', this.onUnhandledRejection);
      delete this.onUnhandledRejection;
    }
  }

  get state() {
    this.checkState();

    if (this._state !== states.HAPPY && this.options.logOnUnhappy && this.options.logger) {
      // log ALL READS if they are a non-happy state to help with debugging
      this.options.logger[this._state === states.UNHAPPY ? 'error' : 'warn'].call(this.options.logger, `[happy-feet] state returned '${this._state}'`);
    }

    return this._state;
  }

  set state(newState) {
    this.updateState(newState, 'State was manually changed', codes.manual);
  }

  get STATE() {
    return states;
  }

  updateState(newState, reason, code) {
    // No change if:
    // 1. we are already in the new state
    // 2. we are in a terminal state (UNHAPPY)
    if (this._state === states.UNHAPPY || this._state === newState) return;

    if (this.options.logger) {
      this.options.logger[newState === states.UNHAPPY ? 'error' : 'warn'].call(this.options.logger, `[happy-feet] state changed from '${this._state}' to '${newState}', reason: '${reason}', code: ${code}`);
    }

    this._state = newState;

    this.emit('change', newState, reason, code);
  }

  checkState() {
    if ((Date.now() - this.startTime) < (this.options.gracePeriod * 1000)) return; // not yet hit grace period

    if (this.options.uncaughtExceptionHardLimit && this.uncaughtExceptions >= this.options.uncaughtExceptionHardLimit) {
      this.updateState(states.UNHAPPY, `uncaughtExceptionHardLimit of ${this.options.uncaughtExceptionHardLimit} exceeded to ${this.uncaughtExceptions}`, codes.uncaughtExceptions);
    } else if (this.options.uncaughtExceptionSoftLimit && this.uncaughtExceptions >= this.options.uncaughtExceptionSoftLimit) {
      this.updateState(states.WARN, `uncaughtExceptionSoftLimit of ${this.options.uncaughtExceptionSoftLimit} exceeded to ${this.uncaughtExceptions}`, codes.uncaughtExceptions);
    }

    if (this.options.unhandledRejectionHardLimit && this.unhandledRejections >= this.options.unhandledRejectionHardLimit) {
      this.updateState(states.UNHAPPY, `unhandledRejectionHardLimit of ${this.options.unhandledRejectionHardLimit} exceeded to ${this.unhandledRejections}`, codes.unhandledRejections);
    } else if (this.options.unhandledRejectionSoftLimit && this.unhandledRejections >= this.options.unhandledRejectionSoftLimit) {
      this.updateState(states.WARN, `unhandledRejectionSoftLimit of ${this.options.unhandledRejectionSoftLimit} exceeded to ${this.unhandledRejections}`, codes.unhandledRejections);
    }

    let memUsage;
    if (this.options.rssHardLimit || this.options.rssSoftLimit) {
      memUsage = memUsage || process.memoryUsage();
      if (this.options.rssHardLimit && memUsage.rss >= this.options.rssHardLimit) {
        this.updateState(states.UNHAPPY, `rssHardLimit of ${this.options.rssHardLimit} exceeded to ${memUsage.rss}`, codes.memory);
      } else if (this.options.rssSoftLimit && memUsage.rss >= this.options.rssSoftLimit) {
        this.updateState(states.WARN, `rssSoftLimit of ${this.options.rssSoftLimit} exceeded to ${memUsage.rss}`, codes.memory);
      }
    }

    if (this.options.heapHardLimit || this.options.heapSoftLimit) {
      memUsage = memUsage || process.memoryUsage();
      if (this.options.heapHardLimit && memUsage.heapTotal >= this.options.heapHardLimit) {
        this.updateState(states.UNHAPPY, `heapHardLimit of ${this.options.heapHardLimit} exceeded to ${memUsage.heapTotal}`, codes.memory);
      } else if (this.options.heapSoftLimit && memUsage.heapTotal >= this.options.heapSoftLimit) {
        this.updateState(states.WARN, `heapSoftLimit of ${this.options.heapSoftLimit} exceeded to ${memUsage.heapTotal}`, codes.memory);
      }
    }

    if (this.options.eventLoopHardLimit && this.eventLoop >= this.options.eventLoopHardLimit) {
      this.updateState(states.UNHAPPY, `eventLoopHardLimit of ${this.options.eventLoopHardLimit} exceeded to ${this.eventLoop}`, codes.eventLoop);
    } else if (this.options.eventLoopSoftLimit && this.eventLoop >= this.options.eventLoopSoftLimit) {
      this.updateState(states.WARN, `eventLoopSoftLimit of ${this.options.eventLoopSoftLimit} exceeded to ${this.eventLoop}`, codes.eventLoop);
    }

    // if an escalation has already been triggered
    if (this.escalationTime && (Date.now() - this.escalationTime) >= this.escalationSoftLimit) {
      // time to escalate and bail out
      this.updateState(states.UNHAPPY, `escalationSoftLimit of ${this.escalationSoftLimit} has been reached!`, codes.escalation);
      return;
    } else if (this.escalationSoftLimit && this._state === states.WARN && !this.escalationTime) {
      // begin countdown if WARN was just triggered and escalations are enabled
      this.escalationTime = Date.now();
    }
  }
}
